File: Platform/Workspace/CSharp/Allors.Workspace/Data/Instanceof.cs
// <copyright file="Instanceof.cs" company="Allors bvba">
// Copyright (c) Allors bvba. All rights reserved.
// Licensed under the LGPL license. See LICENSE file in the project root for full license information.
// </copyright>

namespace Allors.Workspace.Data
{
    using System.Collections.Generic;
    using Allors.Protocol.Data;
    using Allors.Workspace.Meta;

    public class Instanceof : IPropertyPredicate
    {
        public string[] Dependencies { get; set; }

        public Instanceof(IComposite objectType = null) => this.ObjectType = objectType;

        public string Parameter { get; set; }

        public IComposite ObjectType { get; set; }

        public IPropertyType PropertyType { get; set; }

        public Predicate ToJson() =>
            new Predicate
            {
                Kind = PredicateKind.Instanceof,
                Dependencies = this.Dependencies,
                ObjectType = this.ObjectType?.Id,
                PropertyType = this.PropertyType?.Id,
            };
    }
}

File: Platform/Workspace/Typescript/workspace/libs/workspace/data/src/Instanceof.ts
import { ObjectType, PropertyType } from '@allors/workspace/meta';

import { ParameterizablePredicate, ParameterizablePredicateArgs } from './ParameterizablePredicate';

export interface InstanceofArgs extends ParameterizablePredicateArgs, Pick<Instanceof, 'propertyType' | 'objectType'> {}

export class Instanceof extends ParameterizablePredicate {
  propertyType: PropertyType;
  instanceObjectType?: ObjectType;

  constructor(propertyType: PropertyType);
  constructor(args: InstanceofArgs);
  constructor(args: InstanceofArgs | PropertyType) {
    super();

    if ((args as PropertyType).objectType) {
      this.propertyType = args as PropertyType;
    } else {
      Object.assign(this, args);
      this.propertyType = (args as InstanceofArgs).propertyType;
    }
  }

  get objectType(): ObjectType {
    return this.propertyType.objectType;
  }

  toJSON(): any {
    return {
      kind: 'Instanceof',
      dependencies: this.dependencies,
      propertytype: this.propertyType.id,
      parameter: this.parameter,
      objecttype: this.instanceObjectType?.id,
    };
  }
}
